### PR TITLE
Add support for HiDPI in TkAgg on Windows

### PR DESCRIPTION
**Pull request by QuLogic**
## PR Summary

At the moment, Tk does not support updating the 'scaling' value when the monitor pixel ratio changes, or the window is moved to a different one. Thus this needs a patch in the C extension to catch that window message and propagate it to Python. Due to thread locking issues, this needs to be done via a polled queue.

~~This is waiting for the pixel ratio refactor, #19126.~~

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [n/a] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by QuLogic** [comment]
I have not tested on macOS, but this may work if Tk supports it. I tried on Linux and it doesn't seem to work, but I don't have a true HiDPI system to test.

File: lib/matplotlib/backend_bases.py
Comment (anntzer) [inline lib/matplotlib/backend_bases.py:2097]: physical should probably be kwonly

**Comment by QuLogic** [comment]
After quite a bit of trial-and-error, I think this now fully supports HiDPI as much as Windows does.

File: src/_c_internal_utils.c
Comment (anntzer) [inline src/_c_internal_utils.c:140]: I intentionally made this a C module rather than a C++ module to avoid needing the casts (see how XOpenDisplay is loaded above), so you can drop the typedefs and just write
```c
BOOL WINAPI (*IsValidDpiAwarenessContextPtr)(DPI_AWARENESS_CONTEXT) =
    GetProcAddress(user32, "IsValidDpiAwarenessContext");
```
and likewise for the other.

File: src/_c_internal_utils.c
Comment (anntzer) [inline src/_c_internal_utils.c:128]: Would vaguely prefer e.g. mpl_SetProcessDpiAwareness_max (and likewise on the Python side) to stick to the underlying api names.

File: src/_c_internal_utils.c
Comment (anntzer) [inline src/_c_internal_utils.c:155]: this could be factored out of the if()

File: lib/matplotlib/backends/_backend_tk.py
Comment (anntzer) [inline lib/matplotlib/backends/_backend_tk.py:416]: perhaps call the synonym wm_frame() instead of frame() here, for better googlability?

File: src/_tkagg.cpp
Comment (anntzer) [inline src/_tkagg.cpp:128]: should this live in backend_tk, or does it rather belong in _c_internal_utils? (I don't know...)

File: src/_tkagg.cpp
Comment (richardsheridan) [inline src/_tkagg.cpp:106]: Consider using `Tcl_QueueEvent` or `Tcl_EvalEx` to queue up a call to a tkinter-registered python function to avoid the polling loop. Failing that, I think this all occurs in the main thread, so you can use a `deque` instead of a `SimpleQueue`.

File: src/_c_internal_utils.c
Comment (QuLogic) [reply src/_c_internal_utils.c:155]: Then it would be outside the `#if`.

File: src/_tkagg.cpp
Comment (QuLogic) [reply src/_tkagg.cpp:106]: It is not possible to call any Tk functions due to the Tkinter-internal Tcl thread lock. I'm fairly certain I tried `Tcl_QueueEvent` as well, but can do so again.

Comment (richardsheridan) [reply src/_tkagg.cpp:106]: I was under the impression that these windowprocs are called within `Tcl_DoOneEvent`, usually here: 

https://github.com/python/cpython/blob/1e9f0933095403b215c2c4a0be7915d034ff7026/Modules/_tkinter.c#L2938

so I was expecting it to hold the `tcl_lock` already. If that's not the case... sorry for the distraction!

File: src/_c_internal_utils.c
Comment (anntzer) [reply src/_c_internal_utils.c:155]: Ugh indeed.  Perhaps the preprocessor directives could avoid introducing braces, even if that means a tiny bit more repetition?

File: src/_c_internal_utils.c
Comment (QuLogic) [reply src/_c_internal_utils.c:140]: That still causes cast warning on Windows.

File: src/_tkagg.cpp
Comment (QuLogic) [reply src/_tkagg.cpp:128]: I put the DPI awareness setup in `_c_internal_utils` because it is not backend-specific (theoretically, a toolkit could support DPI changes, but expect the application to enable it with the API or a manifest, but I have not really checked this). This function is specific to Tk though, and may or may not be portable to other toolkits.

File: src/_c_internal_utils.c
Comment (QuLogic) [reply src/_c_internal_utils.c:155]: Sure, the duplication doesn't really look that bad in the end, actually.

File: src/_tkagg.cpp
Comment (QuLogic) [reply src/_tkagg.cpp:106]: Actually no, I think I tried `Tk_QueueEvent` which does not work because of said locks. Looking at the API for `Tcl_QueueEvent`, I'm not sure how it could be used. It takes a pointer a function to call later, which would still be my code and unable to (un)lock anything in the Tkinter library.

Comment (QuLogic) [reply src/_tkagg.cpp:106]: Note that Tk/Tcl is threaded on Windows, so it doesn't go through that route, but this one:
https://github.com/python/cpython/blob/1e9f0933095403b215c2c4a0be7915d034ff7026/Modules/_tkinter.c#L2931

As an alternative, I tried calling `Tcl_SetVar` on a variable named per-window, and this might work without crashing. I need to do a bit more testing to be sure, since concurrency issues can be heisenbugs.

File: src/_c_internal_utils.c
Comment (anntzer) [reply src/_c_internal_utils.c:140]: Indeed, but (at least for me) casting the return value of GetProcAddress to (void *) fixes that.

File: src/_tkagg.cpp
Comment (QuLogic) [reply src/_tkagg.cpp:112]: OK, I used a Tk variable now; this is nice as it cleans up the queue polling.

I think what happened originally is that I tried the variable approach from Python, and that failed due to the competing locks which I only found out about some time later.

Comment (richardsheridan) [reply src/_tkagg.cpp:112]: I like it, way better than the original and less magical than what I was imagining.

**Comment by jklymak** [comment]
@richardsheridan are you OK with this now?  I can "officially" approve for you if you are...  Thanks!

**Review by richardsheridan — APPROVED** [review]
Didn't know you were waiting on an explicit approval from me! I'm happy with the new way the DPI change is triggered and with the change overall.

**Comment by jklymak** [comment]
.... I don't think we were, but wanted to make sure you still thought it was on the right track.  This still needs another person to approve.

**Comment by anntzer** [comment]
I can't really claim any expertise on the topic, but I can report that on my (non-hidpi) laptop, this results in toolbar icons being clearly too big (by ~30%, just eyeballing it).

**Comment by QuLogic** [comment]
Good point; it's 24 points, but I should probably scale that down to be whatever the equivalent is.

**Pull request merged**
